src/zenml/integrations/aws/orchestrators/sagemaker_orchestrator.py

```diff
@@ -0,0 +1,186 @@
+
+        Returns:
+            The orchestrator run id.
+
+        Raises:
+            RuntimeError: If the run id cannot be read from the environment.
+        """
+        try:
+            return os.environ[ENV_ZENML_SAGEMAKER_RUN_ID]
+        except KeyError:
+            raise RuntimeError(
+                "Unable to read run id from environment variable "
+                f"{ENV_ZENML_SAGEMAKER_RUN_ID}."
+            )
+
+    def prepare_pipeline_deployment(
+        self, deployment: "PipelineDeployment", stack: "Stack"
+    ) -> None:
+        """Build a Docker image and push it to the container registry.
+
+        Args:
+            deployment: The pipeline deployment configuration.
+            stack: The stack on which the pipeline will be deployed.
+        """
+        docker_image_builder = PipelineDockerImageBuilder()
+        repo_digest = docker_image_builder.build_and_push_docker_image(
+            deployment=deployment, stack=stack
+        )
+        deployment.add_extra(ORCHESTRATOR_DOCKER_IMAGE_KEY, repo_digest)
+
+    @property
+    def settings_class(self) -> Optional[Type["BaseSettings"]]:
+        """Settings class for the Sagemaker orchestrator.
+
+        Returns:
+            The settings class.
+        """
+        return SagemakerOrchestratorSettings
+
+    def prepare_or_run_pipeline(
+        self, deployment: "PipelineDeployment", stack: "Stack"
+    ) -> None:
+        """Prepares or runs a pipeline on Sagemaker.
+
+        Args:
+            deployment: The deployment to prepare or run.
+            stack: The stack to run on.
+        """
+        if deployment.schedule:
+            logger.warning(
+                "The Sagemaker Orchestrator currently does not support the "
+                "use of schedules. The `schedule` will be ignored "
+                "and the pipeline will be run immediately."
+            )
+
+        orchestrator_run_name = get_orchestrator_run_name(
+            pipeline_name=deployment.pipeline.name
+        ).replace("_", "-")
+
+        session = sagemaker.Session(default_bucket=self.config.bucket)
+        image_name = deployment.pipeline.extra[ORCHESTRATOR_DOCKER_IMAGE_KEY]
+        execution_role = (
+            self.config.processor_role or sagemaker.get_execution_role()
+        )
+
+        sagemaker_steps = []
+        for step_name, step in deployment.steps.items():
+            command = StepEntrypointConfiguration.get_entrypoint_command()
+            arguments = StepEntrypointConfiguration.get_entrypoint_arguments(
+                step_name=step_name
+            )
+            entrypoint = command + arguments
+
+            step_settings = cast(
+                SagemakerOrchestratorSettings, self.get_settings(step)
+            )
+            execution_role = step_settings.execution_role or execution_role
+            # accessing value via getattr to avoid mypy error
+            tags = getattr(step_settings, "processor_tags", None)
```

Comment:
What was the mypy issue here? Maybe it's solved by making the attribute non-optional as described in my other comment?

Reply:
Solved with the non-optional attributes changes mentioned above. Problem was Python 3.7 and 3.8 were failing to initialise tests with a dot-notation for `step_settings.processor_tags`, yet when I changed to `step_settings['processor_tags']` then mypy complained that the `SagemakerOrchestratorSettings` object was not indexable. (`Value of type "SagemakerOrchestratorSettings" is not indexable  [index]`). The fix above was the only thing that appeased all parties, but now it's a moot point.